tools/filetools.go

```diff
@@ -121,3 +126,250 @@ func VerifyFileHash(oid, path string) error {
+func FastWalk(dir string, includePaths, excludePaths []string) (<-chan FastWalkInfo, <-chan error) {
+	return FastWalkWithExcludeFiles(dir, "", includePaths, excludePaths)
+}
+
+// FastWalkWithExcludeFiles is like FastWalk but with the additional option to
+// load any file named excludeFilename in any directory, and add its contents
+// to the excludePaths list for that directory and children.
+func FastWalkWithExcludeFiles(dir, excludeFilename string,
+	includePaths, excludePaths []string) (<-chan FastWalkInfo, <-chan error) {
+	fiChan := make(chan FastWalkInfo, 256)
+	errChan := make(chan error, 10)
+
+	go fastWalkFromRoot(dir, excludeFilename, includePaths, excludePaths, fiChan, errChan)
+
+	return fiChan, errChan
+}
+
+// FastWalkGitRepo behaves like FastWalkWithExcludeFiles, preconfigured to ignore
+// the git repo itself (.git) and to load exclude patterns from .gitignore
+func FastWalkGitRepo(dir string) (<-chan FastWalkInfo, <-chan error) {
+	// Ignore all git metadata including subrepos
+	excludePaths := []string{".git", filepath.Join("**", ".git")}
+	return FastWalkWithExcludeFiles(dir, ".gitignore", nil, excludePaths)
+}
+
+func fastWalkFromRoot(dir string, excludeFilename string,
+	includePaths, excludePaths []string, fiChan chan<- FastWalkInfo, errChan chan<- error) {
+
+	dirFi, err := os.Stat(dir)
+	if err != nil {
+		errChan <- err
+		return
+	}
+
+	// This waitgroup will be incremented for each nested goroutine
+	var waitg sync.WaitGroup
+
+	fastWalkItem(filepath.Dir(dir), dirFi, excludeFilename, includePaths, excludePaths, fiChan, errChan, &waitg)
+
+	waitg.Wait()
+	close(fiChan)
+	close(errChan)
+
+}
+
+// Main recursive implementation of fast walk
+// Increment waitg.Add(1) for each new goroutine launched internally
+func fastWalkItem(parentDir string, itemFi os.FileInfo, excludeFilename string,
+	includePaths, excludePaths []string, fiChan chan<- FastWalkInfo, errChan chan<- error,
+	waitg *sync.WaitGroup) {
+
+	fullPath := filepath.Join(parentDir, itemFi.Name())
+
+	if !FilenamePassesIncludeExcludeFilter(fullPath, includePaths, excludePaths) {
+		return
+	}
+
+	fiChan <- FastWalkInfo{ParentDir: parentDir, Info: itemFi}
+
+	if !itemFi.IsDir() {
+		// Nothing more to do if this is not a dir
+		return
+	}
+
+	if len(excludeFilename) > 0 {
+		possibleExcludeFile := filepath.Join(fullPath, excludeFilename)
+		if FileExists(possibleExcludeFile) {
+			var err error
+			excludePaths, err = loadExcludeFilename(possibleExcludeFile, fullPath, excludePaths)
+			if err != nil {
+				errChan <- err
+			}
+		}
+	}
+
+	// The absolute optimal way to scan would be File.Readdirnames but we
+	// still need the Stat() to know whether something is a dir, so use
+	// File.Readdir instead. Means we can provide os.FileInfo to callers like
+	// filepath.Walk as a bonus.
+	df, err := os.Open(fullPath)
+	if err != nil {
+		errChan <- err
+		return
+	}
+	defer df.Close()
+	jobSize := 100
```

Comment:
This variable is effectively a constant. Would you mind extracting it and giving it a name outside of this method?